server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java

```diff
@@ -306,27 +305,11 @@ public static void registerRequestHandler(TransportService transportService, Sea
             (in) -> TransportResponse.Empty.INSTANCE);
 
         transportService.registerRequestHandler(DFS_ACTION_NAME, ThreadPool.Names.SAME, ShardSearchRequest::new,
-            (request, channel, task) -> {
-                searchService.executeDfsPhase(request, (SearchShardTask) task, new ActionListener<SearchPhaseResult>() {
-                    @Override
-                    public void onResponse(SearchPhaseResult searchPhaseResult) {
-                        try {
-                            channel.sendResponse(searchPhaseResult);
-                        } catch (IOException e) {
-                            throw new UncheckedIOException(e);
-                        }
-                    }
-
-                    @Override
-                    public void onFailure(Exception e) {
-                        try {
-                            channel.sendResponse(e);
-                        } catch (IOException e1) {
-                            throw new UncheckedIOException(e1);
-                        }
-                    }
-                });
-            });
+            (request, channel, task) ->
+                searchService.executeDfsPhase(request, (SearchShardTask) task,
+                    new ChannelActionListener<>(channel, DFS_ACTION_NAME, request))
```

Comment:
`ChannelActionListener` also has this weird double-sending logic.

Reply:
Yes, but the code used to propagate the exception out. This seemed to be a left-over from when the `ChannelActionListener` was introduced. The old `map` exception handling would ensure `onFailure` were called on exception. This means this change is effectively a no-op now (with the caveat that `onFailure` will be called on exceptions like for all other `ChannelActionListener` usages).

We should notice that `DirectTransportChannel` will not bubble out exceptions from invoking the `TransportResponseHandler`. So it seems likely that the primary exceptions bubbled out are related to communicating the response over a wire, in which case invoking `onFailure` might be desirable (for instance an NPE on serialization).

So I would like to keep using `ChannelActionListener` here and then deal with `ChannelActionListener` in a follow-up. WDYT?

Reply:
++